### ISSUE TITLE
[Bounty $1,500] ttnn/normalization: sharded softmax accepts a subblock_w larger than Dest and silently returns wrong results

### ISSUE THREAD

**Issue by kinginu** [tenstorrent/tt-metal#52040]
**Component / Area:** ttnn/normalization/softmax

**Issue Type:** Bad Outputs

## Observed

On sentence_bert's own shape and program config, with `subblock_w` raised from 6 to 12, sharded softmax returns rows summing to as much as **47.42** where a softmax must sum to 1. There is no error, no warning and no log line: the op returns a tensor of the right shape and dtype containing garbage. The only bound the op applies to `subblock_w` is `block_w % subblock_w == 0`, and `block_w` is 12 here, so 12 passes it.

The repro below is that measurement. The tables here are the same sweep run one config per process, with a torch PCC alongside the row sums, at `(1, 1, 36864, 384)` `bfloat16`: same shard shape, same 6x8 grid, same `block_h` 24 and `block_w` 12, same per-core work. Default compute config, Blackhole P300, `992f3ca634`:

| subblock_w | row sum mean | min | max | PCC vs torch | |
|---|---|---|---|---|---|
| 1 | 0.999102 | 0.9779 | 1.0192 | 0.999968 | ok |
| 2 | 0.999102 | 0.9779 | 1.0192 | 0.999968 | ok |
| 3 | 0.999102 | 0.9779 | 1.0192 | 0.999968 | ok |
| 4 | 0.999102 | 0.9779 | 1.0192 | 0.999968 | ok |
| 6 | 0.999102 | 0.9779 | 1.0192 | 0.999968 | ok |
| 12 | 6.402125 | 0.8213 | 47.6707 | **0.018846** | wrong |

The threshold moves with the compute config: `subblock_w` tiles are live in Dest at once, and Dest holds 8 tiles by default, 4 with `fp32_dest_acc_en` and 16 with `dst_full_sync_en`. So the same `subblock_w` can be correct under one compute config and garbage under another. `fp32_dest_acc_en=True`, everything else unchanged:

| subblock_w | row sum mean | min | max | PCC vs torch | |
|---|---|---|---|---|---|
| 1 | 0.999166 | 0.9968 | 1.0001 | 0.999997 | ok |
| 2 | 0.999166 | 0.9968 | 1.0001 | 0.999997 | ok |
| 3 | 0.999166 | 0.9968 | 1.0001 | 0.999997 | ok |
| 4 | 0.999166 | 0.9968 | 1.0001 | 0.999997 | ok |
| 6 | 5.043384 | **-1526.7466** | 838.0242 | -0.030736 | wrong, and correct at 16 bits |
| 12 | 7.276325 | 1.9015 | 47.5277 | 0.002739 | wrong |

`block_w = 16` puts a `subblock_w` on both sides of the 8-tile and 16-tile capacities. `dst_full_sync_en` moves the limit the other way:

| subblock_w | dst_full_sync_en=False | dst_full_sync_en=True |
|---|---|---|
| 1 | 0.997798, PCC 0.999962 | 0.997798, PCC 0.999962 |
| 2 | 0.997798, PCC 0.999962 | 0.997798, PCC 0.999962 |
| 4 | 0.997798, PCC 0.999962 | 0.997798, PCC 0.999962 |
| 8 | 0.997798, PCC 0.999962 | 0.997798, PCC 0.999962 |
| 16 | 5.168172, PCC 0.018633, wrong | **0.997798, PCC 0.999962, correct** |

Ruled out:

* Not `numeric_stable`. With `numeric_stable=False`, `subblock_w` 1 to 6 give 0.999680 and `subblock_w` 12 gives row sums from -46.4514 to 49.8968, PCC 0.004103. Both paths fail at the same `subblock_w`.
* Not bfloat16 precision. The same input at `subblock_w` 6 sums to 0.999102 with PCC 0.999968. At 12 the PCC is 0.018846, so the output is uncorrelated with the answer rather than imprecise.
* Not a malformed shard config. `block_h` and `block_w` match the shard shape in tiles, which the op has validated since #37329, and `block_w % subblock_w == 0` holds for every row above.
* Not program cache reuse. Every row above ran in its own process against a freshly opened device.

## Expected

An out-of-range `subblock_w` should raise. The sibling normalization op already does exactly this: `ttnn/cpp/ttnn/operations/normalization/layernorm/device/sharded_layernorm_factory_helpers.cpp:23`, `assert_subblock_compute_config_compatible(dst_full_sync_en, fp32_dest_acc_en, subblock_wt)`, added for #14222, "Make layernorm assert instead of silently giving garbage output when subblock_w too large". Sharded softmax has the same parameter and the same failure and never got the same check.

What should not change: every `subblock_w` at or below the capacity keeps working, including all six in-tree call sites below, two of which sit exactly on the limit. The check has to read the compute config rather than assume 8, or `subblock_w = 16` under `dst_full_sync_en` starts raising even though it is correct today.

## Steps to reproduce

```python
import torch, ttnn

d = ttnn.open_device(device_id=0, l1_small_size=79104)
x = torch.randn((1, 96, 384, 384), dtype=torch.float32).to(torch.bfloat16)
mem = ttnn.MemoryConfig(
    memory_layout=ttnn.TensorMemoryLayout.HEIGHT_SHARDED, buffer_type=ttnn.BufferType.L1,
    shard_spec=ttnn.ShardSpec(
        ttnn.CoreRangeSet({ttnn.CoreRange(ttnn.CoreCoord(0, 0), ttnn.CoreCoord(5, 7))}),
        [24 * 32, 384], ttnn.ShardOrientation.ROW_MAJOR))

for sw in (1, 2, 3, 4, 6, 12):
    pc = ttnn.SoftmaxShardedMultiCoreProgramConfig(
        compute_with_storage_grid_size=(6, 8), subblock_w=sw, block_h=24, block_w=12)
    t = ttnn.from_torch(x, device=d, layout=ttnn.Layout.TILE, memory_config=mem)
    o = ttnn.to_torch(ttnn.from_device(
        ttnn.softmax_in_place(t, program_config=pc, numeric_stable=True))).float().sum(-1)
    print(f"subblock_w={sw:2d}  row sum mean {float(o.mean()):.4f}  min {float(o.min()):.4f}  max {float(o.max()):.4f}")
ttnn.close_device(d)
```

Blackhole P300, `992f3ca634`, nothing applied:

```
subblock_w= 1  row sum mean 0.9930  min 0.9521  max 1.0181
subblock_w= 2  row sum mean 0.9930  min 0.9521  max 1.0181
subblock_w= 3  row sum mean 0.9930  min 0.9521  max 1.0181
subblock_w= 4  row sum mean 0.9930  min 0.9521  max 1.0181
subblock_w= 6  row sum mean 0.9930  min 0.9521  max 1.0181
subblock_w=12  row sum mean 6.4043  min 0.7300  max 47.4248
```

The reading needs no reference implementation: a softmax's rows sum to 1 by definition, so the op reports its own error. The row sums differ in the last digits from the first table because the logical shape is split differently across the same shards, and the third table uses `(1, 1, 6144, 512)` so that `block_w` can be 16.

## Affected call sites

Six in tree at `992f3ca634`, two of them shipped model code and four of them tests. None passes a `subblock_w` above the capacity, so this is a footgun with no current victims rather than a live wrong answer in a shipped model. Two sit exactly on the limit, which is what forces the bound to be `<=` rather than `<`.

| call site | subblock_w | capacity | |
|---|---|---|---|
| `models/demos/blackhole/sentence_bert/ttnn/common.py:70` | 6 | 8 | ok at 16 bits, and wrong under `fp32_dest_acc_en` |
| `models/demos/falcon7b_common/tt/falcon_attention.py:409` | 8 | 8 | on the limit |

The tests are the same story: `tests/ttnn/nightly/unit_tests/operations/fused/test_softmax_sharded.py:72,164,262,344`, `tests/ttnn/unit_tests/operations/fused/test_softmax.py:215`, and `tests/tt_eager/python_api_testing/sweep_tests/pytests/tt_dnn/test_softmax_sharded.py:112`. None exceeds the capacity and none tests above it, which is how this survived. What they do show is that the cap is folklore every caller re-derives by hand, including the `fp32_dest_acc_en` halving that softmax applies by default whenever the input is `float32` (`softmax_device_operation.cpp:386`):

```python
    # calculate the max subblock size < 8
    upper_limit = min(block_w // 2, 8)
```

## Trade-offs a fix has to answer

The capacity depends on the compute config, so a hardcoded `<= 8` would reject `subblock_w = 16` under `dst_full_sync_en`, measured correct above, and would accept `subblock_w = 6` under `fp32_dest_acc_en`, measured wrong above. `ttnn::get_dest_reg_count` already derives the number from both flags and from the tile shape.

Cost on the dispatch path is zero: the values are ones the op already has, and `validate_on_program_cache_miss` runs once per program cache miss rather than per invocation.

Nothing in tree starts raising. Every configuration measured above that is correct today returns bit-identical numbers with the check in place.

## Environment information

* OS: Ubuntu 22.04
* Version of software: tt-metal `992f3ca634`, clean build, nothing applied
* Hardware: Blackhole P300, device 0

Not tested on Wormhole: no hardware available.

**Comment by kinginu** [comment]
I have a patch and tests for this and will open a PR shortly.

**Comment by jberkowitzTT** [comment]
## 📝 Background

This is a community-suggested bounty that we have reviewed & assigned to @kinginu 


## 💡 Problem to Solve

>sharded softmax accepts a subblock_w larger than Dest and silently returns wrong


Note: This bounty is beholden to the [Tenstorrent Bounty Program Terms & Conditions](https://docs.tenstorrent.com/bounty_terms.html) & local tax requirements may apply

**Comment by zhong-math** [comment]
Hi, I'd also like to express interest in this bounty in case the current assignee is no longer active (no PR appears to have been opened since the assignment on Aug 4).

**Understanding of the issue:** the sharded softmax path accepts a subblock configuration that breaks the reduction semantics — softmax along the last dim requires a full-row reduction, so a subblock that doesn't cover the row either produces wrong normalization or requires the kernel to accumulate partial max/sum across subblocks.

**Plan:**
1. Write a host-side reproducer showing where the sharded+subblock configuration diverges from the reference softmax
2. Fix the validation (reject invalid subblock configs with a clear error) and/or fix the kernel's partial-reduction handling
3. Add regression tests for valid/invalid subblock shapes

Standing by in case this frees up. Thanks!

**Comment by dickwin2003** [comment]
Hi @jberkowitzTT,

I'd like to be considered for this bounty if the current assignment has gone inactive. It was claimed on Aug 4 and there is still no PR (~5 weeks); the bounty terms allow reassignment after two weeks of unresponsiveness.

If @kinginu is still on it, please just let me know — I'll gladly stand down.

This one is well suited to a host-only approach, and I can move quickly: the repro sweep in the issue already pins the failure (subblock_w=12 accepted on block_w=12 with fewer Dest rows available, silently returning row sums around 47). My deliverables: (1) a host-side validation/repro script, (2) the fix — validating subblock_w against available Dest rows at program-config time and erroring loudly instead of silently corrupting output, (3) regression tests covering the sweep shapes. Background: numerics debugging and kernel test pipelines, currently building edge inference on Windows + Vulkan iGPU.

Thanks!

**Comment by dickwin2003** [comment]
Correction to my comment from yesterday: I said there was no PR, but I missed that @kinginu’s PR #54388 has been open since Aug 26 and is CI-green — my mistake, and the bounty terms make the existing assignment decisive anyway. Withdrawing my request to take this over; best of luck with the review on #54388. Apologies for the noise.

**Comment by sjz207** [comment]
Submission for this bounty: **PR #56163** — two `subblock_w` guards in the sharded path (`> 0` + the Dest-capacity bound via `get_dest_reg_count`).

*This is a submission by an AI agent. Wallet (Base USDC): `0xDeB7a2C40E4BAe45f2427d5AF410C56F2AC5E4DC`*

**Comment by jberkowitzTT** [comment]
Hi @kinginu - congratulations on completing this bounty, and thank you for the great work!

To initiate the payment process, please email us at bounties@tenstorrent.com with a link to this issue

**Comment by jberkowitzTT** [comment]
Hi @kinginu- congratulations on completing this bounty, and thank you for the great work!

To initiate the payment process, please email us at bounties@tenstorrent.com with a link to this issue
